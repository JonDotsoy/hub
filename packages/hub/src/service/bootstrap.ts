import { Router } from "artur";
import { Hub } from "../hub/hub";
import { settings } from "./settings";
import * as YAML from "yaml";
import {
  DynamicContentResponse,
  dynamicContentMiddleware,
} from "./utils/dynamic-content-response";
import { protoGrpcType, serviceProto } from "./service-proto-server";
import type { HubServiceHandlers } from "./protos/HubService";
import type {
  ServiceDefinition,
  UntypedServiceImplementation,
} from "@grpc/grpc-js";

const openapi = new URL("./openapi.yml", import.meta.url);

const protoText = await Bun.file(serviceProto).text();

export type ProtoServiceDefinition = {
  service: ServiceDefinition;
  handlers: UntypedServiceImplementation;
};

export const bootstrap = async () => {
  const router = new Router();
  const hub = await Hub.from(settings.hubSchema);

  const protoServiceDefinitions: ProtoServiceDefinition[] = [
    {
      service: protoGrpcType.HubService.service,
      handlers: {
        isAllowed: (call, callback) => {
          hub
            .isAllowed(call.request)
            .then((allowed) => {
              callback(null, { allowed });
            })
            .catch((err) => {
              callback(err, null);
            });
        },
      } satisfies HubServiceHandlers,
    },
  ];

  router.use("GET", "/proto", {
    fetch: (req) => {
      return new Response(protoText, {
        headers: {
          "content-type": "application/protobuf",
        },
      });
    },
  });

  router.use("POST", "/is-allowed", {
    fetch: async (req) => {
      const body = await req.json();

      const allowed = await hub.isAllowed({
        principalId: body.principalId,
        resource: body.resource,
        action: body.action,
      });

      return Response.json({ allowed });
    },
  });

  router.use("GET", "/", {
    middlewares: [dynamicContentMiddleware],
    fetch: (req) =>
      new DynamicContentResponse({
        uptimeMs: Math.floor(process.uptime()),
        apis: router.routes
          .map((route) => {
            const relativeUrl = new URL(route.urlPattern.pathname, req.url);

            if (!route.urlPattern.test(relativeUrl)) return [];

            return {
              method: route.method,
              url: relativeUrl,
            };
          })
          .flat(),
      }),
  });

  router.use("GET", "/openapi", {
    middlewares: [dynamicContentMiddleware],
    fetch: async (req) => {
      const openapiPayload = YAML.parse(await Bun.file(openapi).text());

      openapiPayload.servers = [
        {
          url: new URL(settings.base, settings.site ?? req.url).toString(),
        },
        ...(openapiPayload.servers ?? []),
      ];

      return new DynamicContentResponse(openapiPayload);
    },
  });

  router.use("GET", "/manifest", {
    middlewares: [dynamicContentMiddleware],
    fetch: async (req) => {
      return new DynamicContentResponse(hub.toJSON());
    },
  });

  return {
    httpRouter: router,
    protoServiceDefinitions,
  };
};
