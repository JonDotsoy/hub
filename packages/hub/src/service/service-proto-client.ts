import grpc, {
  type ServiceClientConstructor,
  type ServiceDefinition,
} from "@grpc/grpc-js";
import protoLoader from "@grpc/proto-loader";
import type { ProtoGrpcType } from "./protos/service";

export const serviceProto = new URL("./protos/service.proto", import.meta.url);

const packageDefinition = await protoLoader.load(serviceProto.pathname, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true,
});
const routeguide: any = grpc.loadPackageDefinition(packageDefinition);
// console.log("ðŸš€ ~ routeguide:", new routeguide.HubService)
export const HubService: ProtoGrpcType["HubService"] = routeguide.HubService;
// const e = ('RouteGuide' in routeguide) ? new routeguide.RouteGuide() : null;
