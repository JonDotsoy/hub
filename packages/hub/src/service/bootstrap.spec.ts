import { describe, it, expect } from "bun:test";
import { bootstrap, type ProtoServiceDefinition } from "./bootstrap";
import type { Router } from "artur";
import grpc from "@grpc/grpc-js";
import { HubService } from "./service-proto-client";
import type { Validation__Output } from "./protos/Validation";

async function useServer(
  httpRouter: Router,
  protoServiceDefinitions: ProtoServiceDefinition[],
) {
  const httpServer = Bun.serve({
    port: 56987,
    fetch: async (req) =>
      (await httpRouter.fetch(req)) ?? new Response(null, { status: 404 }),
  });
  const protoServer = new grpc.Server();

  for (const { service, handlers } of protoServiceDefinitions) {
    protoServer.addService(service, handlers);
  }

  await new Promise((resolve, reject) => {
    protoServer.bindAsync(
      "0.0.0.0:56988",
      grpc.ServerCredentials.createInsecure(),
      (err, port) => {
        if (err) return reject(err);
        return resolve(port);
      },
    );
  });

  return {
    get httpURL() {
      return httpServer.url;
    },
    get protoHost() {
      return new URL("proto://localhost:56988").host;
    },
    [Symbol.dispose]() {
      httpServer.stop();
      protoServer.forceShutdown();
    },
  };
}

describe("bootstrap", () => {
  it("serves the proto definitions", async () => {
    const { httpRouter, protoServiceDefinitions } = await bootstrap();
    using server = await useServer(httpRouter, protoServiceDefinitions);
    const res = await fetch(new URL("/proto", server.httpURL));
    expect(res.status).toEqual(200);
    expect(res.headers.get("content-type")).toEqual("application/protobuf");
    expect(await res.text()).toMatchSnapshot();
  });

  it("calls the hub service", async () => {
    const { httpRouter, protoServiceDefinitions } = await bootstrap();
    using server = await useServer(httpRouter, protoServiceDefinitions);

    const hubService = new HubService(
      server.protoHost,
      grpc.credentials.createInsecure(),
    );

    const value = await new Promise<Validation__Output | undefined>(
      (resolve, reject) => {
        hubService.isAllowed(
          { principalId: "a", action: "b" },
          (err, value) => {
            if (err) return reject(err);
            return resolve(value);
          },
        );
      },
    );

    expect(value?.allowed).toBeFalse();
  });
});
