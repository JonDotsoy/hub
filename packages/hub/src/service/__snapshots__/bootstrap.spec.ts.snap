// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`bootstrap serves the proto definitions 1`] = `
"syntax = "proto3";

service HubService {
    rpc isAllowed (Body) returns (Validation) {}
}

message Body {
    string principalId = 1;
    optional string resourceJson = 2;
    string action = 3;
}

message Validation {
    bool allowed = 1;
}
"
`;
